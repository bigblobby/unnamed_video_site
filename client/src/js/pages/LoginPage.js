import React, { useState } from 'react';
import { useDispatch } from "react-redux";
import {userLogin} from "../actions/userActions";

function LoginPage(props){
    const [ username, setUsername ] = useState('');
    const [ password, setPassword ] = useState('');
    const dispatch = useDispatch();
    const previousUrl = props.location.state?.from.pathname;

    function handleSubmit(e){
        e.preventDefault();

        dispatch(userLogin({
            username: username,
            password: password
        }, previousUrl));
    }

    return (
        <div>
            <h1>The login page</h1>

            <form onSubmit={handleSubmit}>
                <label htmlFor="username">Username:</label>
                <input type="text" name="username" id="username" onChange={ (e) => setUsername(e.target.value) }/>
                <label htmlFor="password">Password:</label>
                <input type="password" name="password" id="password" onChange={ (e) => setPassword(e.target.value) }/>

                <button type="submit">submit</button>
            </form>
        </div>
    )
}

export default LoginPage;
