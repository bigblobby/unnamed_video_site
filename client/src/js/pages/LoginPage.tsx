import React from 'react';
import { connect, useDispatch } from "react-redux";
import { userLogin } from "../actions/userActions";

type LoginPageState = {
    username: string;
    password: string;
    previousUrl: string;
}

type LoginPageProps = {
    userLogin: (user, url) => {}
}


class LoginPage extends React.Component<LoginPageProps, LoginPageState> {
    constructor(props) {
        super(props);

        this.state = {
            username: '',
            password: '',
            previousUrl: props.location?.state?.from.pathname
        }
    }

    handleSubmit = (e) => {
        e.preventDefault();

        this.props.userLogin({
            username: this.state.username,
            password: this.state.password
        }, this.state.previousUrl)
    };

    handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({
            [e.target.name]: e.target.value
        } as { [K in keyof LoginPageState]: LoginPageState[K] } );
    }

    render(){
        return (
            <div>
                <h1>The login page</h1>

                <form onSubmit={this.handleSubmit}>
                    <label htmlFor="username">Username:</label>
                    <input type="text" name="username" id="username" onChange={ this.handleChange }/>
                    <label htmlFor="password">Password:</label>
                    <input type="password" name="password" id="password" onChange={ this.handleChange }/>

                    <button type="submit">submit</button>
                </form>
            </div>
        )
    }
}

const mapStateToProps = ({userReducer}) => {
    const {user, error} = userReducer;
    return {
        user,
        error
    }
};

const mapDispatchToProps = { userLogin };

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);
