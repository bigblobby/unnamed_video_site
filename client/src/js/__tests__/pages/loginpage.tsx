import React from "react";
import {render, screen} from "../test-utils";
import '@testing-library/jest-dom/extend-expect'
import userEvent from '@testing-library/user-event'
import LoginPage from "../../pages/LoginPage";
import Api from '../../services/ApiService';


describe("LoginPage component", () => {
    test("should match snapshot", () => {
        const { asFragment } = render(<LoginPage />)
        expect(asFragment()).toMatchSnapshot();
    });

    test("should have a username field", () => {
        render(<LoginPage />)
        expect(screen.getByLabelText('Username:')).toBeInTheDocument();
    });

    test("should have a password field", () => {
        render(<LoginPage />)
        expect(screen.getByLabelText('Password:')).toBeInTheDocument();
    });

    test("username insert text", () => {
        render(<LoginPage />)

        userEvent.type(screen.getByLabelText('Username:'), 'dave');
        expect(screen.getByLabelText('Username:')).toHaveValue('dave');
    });

    test("password insert text", () => {
        render(<LoginPage />)

        userEvent.type(screen.getByLabelText('Password:'), 'somepassword');
        expect(screen.getByLabelText('Password:')).toHaveValue('somepassword');
    });

    test("log user in successfully", (done) => {
        const fakeUserResponse = {
            message: 'success',
            token: 'fake_user_token',
            user: {}
        }

        // @ts-ignore
        jest.spyOn(Api, 'loginUser').mockImplementationOnce(() => {
            return Promise.resolve(fakeUserResponse)
        });

        render(<LoginPage />)

        userEvent.type(screen.getByLabelText('Username:'), 'dave');
        userEvent.type(screen.getByLabelText('Password:'), 'somepassword');
        userEvent.click(screen.getByText('submit'));

        setImmediate(() => {
            expect(window.localStorage.getItem('user_token')).toEqual(fakeUserResponse.token)
            done();
        });
    });
});
