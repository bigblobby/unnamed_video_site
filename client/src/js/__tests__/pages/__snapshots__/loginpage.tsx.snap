// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage component should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      The login page
    </h1>
    <form>
      <label
        for="username"
      >
        Username:
      </label>
      <input
        id="username"
        name="username"
        type="text"
      />
      <label
        for="password"
      >
        Password:
      </label>
      <input
        id="password"
        name="password"
        type="password"
      />
      <button
        type="submit"
      >
        submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
